restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java

```diff
@@ -660,6 +668,7 @@ private void writeNewArrayState(
         bindParams,
         processedPath,
         timeSource.currentTimeMicros(),
+        ttl,
```

Comment:
What I was thinking was the following situation:

1. I add a document with TTL
2. Then I patch same document at some point (not providing TTL, because how would I calculate now what should it be
3. When TTL is over part inserted in step one gets deleted, but patched part is still there..

Reply:
The TTL value is sticky - If you set it and then don't touch it, it will not be overwritten.

So for your case, in 1. you would set it, and since in 2. you are leaving it `null` (and not re-setting it), it will be left alone, so 3. would not occur.

This is vaguely mentioned here (you have to set it explicitly to 0 to "cancel" the TTL): https://docs.datastax.com/en/dse/5.1/cql/cql/cql_using/useExpireExample.html

Reply:
I think this is not true.. Let's say at step (1) I am inserting a doc and that adds 10 rows, all will have TTL. In the step (2) I  patch the document and add 5 rows without the TTL. In step (3) I am left with only 5 rows from the patch, right?

It's easier when we have a single row insert and update, but here we work against docs and they result in multiple rows. That's what I had in mind.

Reply:
oh never mind - there is an issue here because we are deleting a whole set of rows and then adding an entire different set of rows...yea I'm not sure how to reconcile this one, other than to select out the current TTL of the rest of the document and then explicitly set to that value.

That kind of ruins read-before-write for our PATCH/PUT operations though...I'll have to think on this a bit more

Reply:
Yes, but I think we can not solve this so easily.. We can say that's a limitation, however, I guess this is something users would not be happy about..

Reply:
Yeah this is no good as a limitation because if you set TTL on a document and then PATCH without a TTL, you have no way of getting it to behave in a "default" or reasonable manner - the chunk of document you patch with will persist forever and there's no way around it...also there's no way to get it to match the current TTL of the rest of the document without read-before-write. Maybe I need to add a parameter to make the patch do a slower operation (a read first) in order to attempt to match the TTL of the document that exists...

Reply:
What behavior is desirable / expected here? From a user perspective, I'd expect TTL to apply at the document level, not that portions of the document could expire at different times. I'm not sure about the idea of updating TTL on an existing document, whether that would be a good idea or not.

Reply:
@jeffreyscarpenter the expected behavior would be that if you set a TTL on a newly created document, and then you patch a sub-path of the document, the patched sub-path should have the same TTL as the document.

Currently if you update a document that has a TTL, the patched part would have no TTL (and therefore would unintentionally last forever) - making it automatically try to match the surrounding document's TTL would require read-before-write, which would impact performance every time (even when the document has no TTL, because you would have to check the database to see if the document has any TTL)

So the only way to implement this I think would be slightly more un-ergonomic because documents are really more than one row...and maybe every time you patch a document that you know has a TTL, you have to supply `&ttl=match`, in order to do the more expensive read-before-write and match the TTL of the document. And if you don't supply it, the sub-document will last forever (no TTL).

I'm open really to any other suggestion, this is surprisingly complex for what I thought was going to be a straightforward impl

Reply:
Maybe I will call it `ttl=auto` to make it a lil clearer

Reply:
Was this issue considered when document shredding idea was first conceived? And if so, what were the expected semantics? It really sounds like a very complicated issue at low level, even if I agree that from user perspective it's most logical that the whole document has same TTL settings.

Or alternatively that one should not have auto-expiring documents at all (but possible allow auto-expiring values, i.e. per row)?

Reply:
It was definitely not considered to begin with :/ but since Cassandra allows TTL on rows, a user that is coming from cassandra to docs API currently wants this feature...

Reply:
In my opinion, there should only be one TTL for a document. This is kind of an advanced feature and if someone is opting in to use it, they should do so knowing that they are asking for a read-before-write. Would it help if the TTL was stored as document metadata?

Reply:
Conceptually I agree but I am not sure how viable this is since I don't know if:

1. there actually is separate per-document metadata (I thought it's just a bunch of rows (each representing a leaf node))
2. underlying storage will operate on per-row (or per-field-value) basis and has no concept of "this set of rows work logically as a bigger entity"

Now obviously one can, with lots of work and overhead, work around this by synchronizing all rows. But it seems fragile and possibly a losing battle. It may lead to the worst kinds of flaws -- things are very rare, but with enough traffic, not rare enough. So we could get cases where arbitrary document nodes just disappear due to TTL value being off, having to investigate, and without much to go on.

So I hope we are very confident that our desire to make TTL work for Documents does not come to bite us in near future.

Does Mongo support TTL for documents or parts therein?